### ISSUE TITLE
85A module not working.

### ISSUE THREAD

**Issue by MickTheMechanic** [iliasam/Laser_tape_reverse_engineering#5]
In have purchased the following device, which appears to be identical to the D30 device:
https://www.amazon.de/gp/product/B08H8WPB3M/ref=ppx_yo_dt_b_asin_title_o03_s00?ie=UTF8&psc=1

Upon disassembly i noticed the module was marked as 85A, but it appears to be visually identical to the 85B module:
![Image (20)](https://user-images.githubusercontent.com/44751666/140650147-b861d287-e3fa-4d66-8f01-2a6e45e3aa77.jpeg)
![Image (19)](https://user-images.githubusercontent.com/44751666/140650150-03f74479-598a-4bf5-8a98-f38bb6dd6302.jpeg)

On the JRT-Measure website it seems like the 85A is the same as the 85B, just with a more powerful laser diode, better suited for industry applications:
https://www.jrt-measure.com/distance-sensor-short-range/57266349.html

I am able to succesfully program the MCu via the STLink software using the U85B HEX files, but I do not get any data via the serial connection, and the laser is permenantly on.

I have tried using an arduino and a CH340 shield to interface with my PC, and I have tried the serial monitor in the Arduino IDE, Serial Port Monitor, and the PC utilities, but I am not getting any comminication from the module.

Could you provide some support for troubleshooting?

**Comment by iliasam** [comment]
What is the name of HEX file that you are using?
Do you use ST-Link programmer or DFU bootloader?

**Comment by MickTheMechanic** [comment]
I have tried STLink programmer and STcube programmer.

I have tried the following files from the CortexM0 folder:
Firmware_dist_calculation_fast_boot.hex
Firmware_dist_calculation_fast.hex
Firmware_dist_calculation_simple.hex
Firmware_raw_capture.hex

Each upload appears to be sucsessful, but no communication via serial. 

It is possible i have made a mistake somewhere, it is my first time working with STM32 products, until now my experience is only with ATMega and ESP32 products.

**Comment by MickTheMechanic** [comment]
I checked the documentation again, I think i have made a mistake.

When using the programmer which file do I select?

**Comment by iliasam** [comment]
Try to use "Firmware_dist_calculation_fast_boot.hex" again with "PC_show_results.exe" utility.
Your USB-UART converter should support baudrate - 256000.
It would be good to check TX pin of the module by oscilloscope.
Also you can try to disconnect beeper from PCB and connect it between TX pin and GND. You should hear buzzing if module is transmitting data (I tried it).

If you are using NOT "_boot.hex" file, you need to connect module UART RX pin to external UART which will pull up this line - to get UART working.
"_boot.hex" files do not have such check , UART is always enabled in such firmware.

**Comment by iliasam** [comment]
Sorry, my mistake BOOT0 must be connected to GND or it can be floating to get firmware running.

**Comment by MickTheMechanic** [comment]
In the STCube software which baudrate should i select for the uplaud via UART?

**Comment by iliasam** [comment]
Baudrate for bootloader is 115200 - you can see at image: https://github.com/iliasam/Laser_tape_reverse_engineering/wiki/U85-module-programming
I have updated text and pinout about BOOT0 pin just now, be careful.

**Comment by MickTheMechanic** [comment]
I cannot connect, i have the error message: 

```

17:07:15 : Serial Port COM4 is successfully opened.
  17:07:15 : Port configuration: parity = even, baudrate = 115200, data-bit = 8,                     stop-bit = 1.0, flow-control = off
  17:07:17 : Timeout error occured while waiting for acknowledgement.
  17:07:17 : Timeout error occured while waiting for acknowledgement.
  17:07:17 : Error: Activating device: KO. Please, verify the boot mode configuration and check the serial port configuration. Reset your device then try again... 
  17:07:21 : RTS low
  17:07:21 : DTR low

```


I currently have PWRON and BOOT0 connected to VBAT, and NRST is floating. Is this correct?

**Comment by iliasam** [comment]
"PWRON and BOOT0 connected to VBAT, and NRST is floating." - this is correct.
So it is strange, that you can't connect from bootloader.
Are you sure that RX and TX are connected OK (not swapped)?

**Comment by MickTheMechanic** [comment]
I have tried both connections, but i cant get the connection to work. Is it possible i have damaged the bootloader somehow?

**Comment by iliasam** [comment]
Strange situation.
st-link programmer with NRST connected and  "Connect under Reset"  selected should we working anyway.
What is the type of st-link programmer that you are using?

**Comment by MickTheMechanic** [comment]
I can connect with the STlink, but not with serial. I am currently re-doing all my connections, incase it was an issue with the cables or soldering.

I have the STLink V2, looks just like a usb thumb drive

**Comment by iliasam** [comment]
If your programmer is same: https://www.mikroshop.ch/shoppic/ST_LINKV2.jpg
You need to notice, that it do not have NRST pin. Its RST pin for STM8 only: https://research.kudelskisecurity.com/2020/04/15/swd-part-3-swo-and-nrst/
But if can erase and flash MCU without NRST connected to st-link, so that is OK.

**Comment by MickTheMechanic** [comment]
It looks like i damaged the module somehow, I have a second one here and I have sucsessfully connected via Serial

**Comment by MickTheMechanic** [comment]
I have done the calibration at >10cm, but the distance is way off. It is showing 2.8m at a distance of only 1.7m. 

Are there further steps that need to be taken to improve accuracy?

**Comment by iliasam** [comment]
Could you try to go calibration at distance 20cm?
I do not have such big errors at my module.

**Comment by MickTheMechanic** [comment]
Ok, i will do some testing and let you know how it goes. 

I noticed some components on the circuit board are different when compared to the U85B, most notibly is the large component (an inductor?) is no longer present, and the 6 pin IC marked FB3MKE has been replaced by 2 smaller 5 pin IC's labelled YJ33. Do you know what this circuit is responsible for? 

Also, do I understand correctly that the _simple file is for precision, _fast file is for speed? 

And a final question for now, i have been observing the signals in the PC utility, and I noticed the low frequency (FREQ LOW) signal does not look clean, whereas the other frequencies are a very clean sinus wave. Is this normal?


![Screenshot (49)](https://user-images.githubusercontent.com/44751666/140656710-c36887b6-38fa-41e9-bda0-142767acefc1.png)
?

**Comment by iliasam** [comment]
FB3MKE IC (DC-DC) is used for producing 3.3V for power supply. YJ33 is a linear source of 3.3V.
I'm afraid that "85a" version could really need other phase compensation method.

"Also, do I understand correctly that the _simple file is for precision, _fast file is for speed?"
No, they have same precision, but "_simple " is just written simpler to understand, but it is slower.

" low frequency (FREQ LOW) signal does not look clean"
That is OK, this "FREQ LOW" and "FREQ5" are experimental.

**Comment by MickTheMechanic** [comment]
Do you mean temperature compensation? 

Another question I have is about the photosensor voltage, according to the pc utility it is at 110v, is this normal? And should this value change or remain constant?

**Comment by iliasam** [comment]
Phase is compensated by measured amplitude value, temperature phase offset is compensated too, but temperature dependence is much smaller.

110v is OK, so be careful with module. Module does not produce really dangerous current, but such voltage hurts and could easily destroy low-voltage electronics if some short circuit happens.  
APD voltage is changed when measured temperature is changing.

**Comment by MickTheMechanic** [comment]
I would like to get this module working correctly, it may also benefit others in the future.

could you walk me through how to determine a new phase compensation for this hardware? I can do testing or take measurements with an oscilloscope, I just don’t know where to begin.

**Comment by iliasam** [comment]
I can try to give you description how to collect needed data.
Firstly, I need to build firmware with compensation disabled - it must be used with "PC_phase_analyse.exe" utility.
You would not need any other electronic devices co collect data.
But you need to have some kind of movable plate that must be installed close to laser output lens.
This plate is needed to change amplitude of transmitted light accurately.  See this photo: https://habrastorage.org/r/w1560/web/992/4ae/5f5/9924ae5f5807467fa40f9595134ac2f4.jpg

**Comment by MickTheMechanic** [comment]
Ok no problem, I can manage some kind of mechanism for that, let me know when you have the modified firmware.

i have read your article in more detail and i have some observations and questions..

with the current firmware I observe that the distance measurement is not correct, one mm measured from the sensor corresponds to multiple centimeters, apart from that the distance seems stable until a distance of about 30cm (95mm according to the sensor) is reached, then the value jumps up to over 1000mm, it is the stable once more for a short distance before jumping again.

I wonder if this phenomenon is caused by the phase difference exceeding 360°, Or perhaps the reference frequency is no longer 5khz with the new hardware? Just some thoughts I had while reading your paper.

also thank you for the support 👍

**Comment by iliasam** [comment]
All frequencies are generated by PLL IC (si5351 or its clone), which is same type in all types of modules. This IC is fully controlled by  MCU, so all frequencies should be OK.

**Comment by MickTheMechanic** [comment]
Ok, thank you for the explanation.

I will begin work on the mechanism for limiting the Laser output.

**Comment by iliasam** [comment]
I have published test firmware: https://github.com/iliasam/Laser_tape_reverse_engineering/blob/master/Code/CortexM0/test_firmware_phase_calc.hex
See readme.txt description: https://github.com/iliasam/Laser_tape_reverse_engineering/tree/master/Code/CortexM0
I need some time to write description how to make measurements.

**Comment by MickTheMechanic** [comment]
Thanks, I have also finished the device for limiting the amplitude:
![AD2A950B-23DF-45BB-BCA0-FF5730317DA3](https://user-images.githubusercontent.com/44751666/140807436-b7bc2354-0af6-41c4-8251-6ee1f13f6629.jpeg)
![67F5A488-969F-42B8-B325-C094F7DA41FF](https://user-images.githubusercontent.com/44751666/140807468-8c5ac478-ee59-4fb7-9551-d2fb384a9915.jpeg)
![C42F64BC-F206-4FB3-9573-08E366FD07E5](https://user-images.githubusercontent.com/44751666/140807476-ed35bd73-374e-449a-aa53-3372b8388b6e.jpeg)

I await your further instructions.

thanks for the help

**Comment by MickTheMechanic** [comment]
In your reply you said the plate should cover the laser output, but in the image it seems like the plate is covering the photo sensor input. Should the input or the output be covered?

**Comment by iliasam** [comment]
Yes, in my photo plate is closing photodiode lens, but photo is made with old module. I had closed laser output with 85b module.
I suppose that your device should work OK. You can test it - load test firmware, run  "PC_phase_analyse.exe" utility and check amplitude at different positions of plate.

**Comment by MickTheMechanic** [comment]
I have tested the device and I can adjust the amplutude by adjusting how much the laser is covered.

What are the next steps?

**Comment by iliasam** [comment]
Needed measurement steps:
1. You need to fix module somehow so that its beam should not move when you are changing plate position.
2. You need to point laser beam to some white object, this object should be movable is range 0-1.5m from module.
Movement direction should be parallel to laser beam (you don't need super high accuracy here).
3. Power on module, open photodiode plate, and wait 5 min to get warm up. Run "PC_phase_analyse.exe" and confirm that values are measured correctly.
You need to control values:  "Amplitude" and "Freq 1 Phase AVR". "AVR" is averaged value, you can see that this value is changing slowly, so you need to wait for a certain time until this value will get stable after any changing of plate position.
4. Move object along laser beam until "Freq1 phase" will be max close to 0.0 deg, align object position to get "Freq 1 Phase AVR" closet to 0.0 deg too. This would be **zero distance position**.
5. Change position of photodiode plate by small steps so amplitude should decrease to ~100 steps.
It would be good to keep amplitude close to x00 values.
So you need to create table like:
Amplitude1000 - phase#1
Amplitude900 - phase#2
...
Amplitude100 - phase#A
Amplitude50 - phase#B
6. You need to move object to 230 mm from **zero distance position** (check distance by ruler) and do same measurements as in step 5.
7. You need to move object to 460 mm from **zero distance position** (check distance by ruler) and do same measurements as in step 5.
Share measured 3 tables.

**Comment by MickTheMechanic** [comment]
[AmplitudeValues.ods](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7523480/AmplitudeValues.ods)

**Comment by MickTheMechanic** [comment]
Is this what you need?

**Comment by iliasam** [comment]
Thanks for data. 
All measurements are made correctly.
Sorry, but I need to say that this data are hard to analyze because it looks like you get signal overdrive.
It would be great, if you can repeat measurements but do small a change in method.
You need to change object (use darker object) so that amplitude value should never be never bigger than 1600.
I suppose that you will get another zero distance position in this way.
Also, zero distance position should be more than 30cm from the module.
You can do increase amplitude steps to 200 to save your time, but you still need to make measurements at 50 and 100 amplitude values.

**Comment by MickTheMechanic** [comment]
I have re-measured the values. I changed the surface from a gloss white to a matt white, and made a new 0 point whaich was 500mm from the sensor.

Here are the new values:
[AmplitudeValues2.ods](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7541600/AmplitudeValues2.ods)

**Comment by iliasam** [comment]
Thank you for data.
It looks really different to mine measurements, so it would require some time to process this data.

**Comment by iliasam** [comment]
Could you do additional test:
Fully open receiver lens.
Place object to a "zero distance" position.
Confirm that "Phase1" is zero, and amplitude is near 1500 (as you measured earlier).
Try to close manually **laser** with something thin (knife is suitable). Check that light is not reflected to receiver lens.
Confirm that phase value is really begin to grow fast and reach 45 deg when amplitude is changing from 500 to 50 points.

**Comment by maillemaker** [comment]
Thanks for working on this.  I have purchased some laser tape measures:

https://www.aliexpress.com/item/1005001772064136.html

And they came with 85A modules.  I would like to integrate a laser rangefinder with my Arduino.

**Comment by MickTheMechanic** [comment]
Unfortunately I have damaged my remaining module. I must order some replacements, this means the project is temporarily on hold.

**Comment by iliasam** [comment]
I'm orded one devise here: https://www.aliexpress.com/item/1005001772064136.html
I would try to investigate it.

**Comment by bdevijve** [comment]
Hello,
I'm following this very interesting project and I own 2x AtuMan LS-1 / LS-1P from Banggood which run 85A as well...

**Comment by MickTheMechanic** [comment]
I have a request for the new firmware, to make the baudrate user selectable by setting a gpio pin high or low.

For example is the pin is grounded then the device runs a lower baudrate, something like 56000, which is more Arduino friendly.

If the pins are floating/High then the device runs a higher baudrate, for using with PC utility.

Currently PA11 and PA12 are unused, and could be set low or high by modifying the button circuitboard included in the range finder.

Alternatively the pin TIM3/Ch2 (Buzzer pin) is also available, and it looks like it can also work as an input pin, which has the advantage of solder pads being located on the device board.

Is this something you would consider with the new release for 85a board?

**Comment by iliasam** [comment]
@MickTheMechanic, please open new issue for baudrate request.

I'm still waiting for my new device with probably 85a module.

**Comment by iliasam** [comment]
I have received my D30 laser rangefinder. It have 85A module.
I had notice that Li-ion battery is directly connected to the module. Battery voltage is usually in range 3.6-4.2V.
You can see two same ICs (with 5 legs) near power pads. Mine are marked "LB33", which can be decoded as MIC5205-3.3BM5.
Pinout is looking suitable. So these are Linear Voltage Regulator ICs, they produce 3.3V for the module.
That means that **input voltage must be higher than 3.3V.**
Also I tested how input voltage is displayed at the screen. All voltages below 3.7V are showed with low battery sign which is correct for Li-ion battery.

All previous modules were designed to be powered from two AAA batteries, so they have step-up DC-DC converter IC, which convert 2.5-3V from batteries to 3.5V. 85B module also have such module, so I'm not sure 85B module can be powered from Li-ion battery. Mine laser tape measure with 85B was powered by two AAA batteries.
Interesting thing: D35 Laser Tape Measure is designed to be powered from AA batteries so it can contain 85B module (of course, I'm not sure in it).

That is why I powered my 85A module with 3.7V.
I have replaced original firmware of 85A with mine own (for 85B module) with bootloader (without using programmer).
After that I have checked how firmware is working - and it is working OK without any modifications.
I tried to change signal amplitude by closing laser beam - it is not affecting the measured values.

**Comment by bdevijve** [comment]
Very good point if the 85B firmware does not need any change for 85A then !
Can you point me to the flashing instructions as you performed ? (With bootloader) and last question : do you think we can power this device with 5V ?

**Comment by iliasam** [comment]
See https://github.com/iliasam/Laser_tape_reverse_engineering/wiki/U85-module-programming.  
See also vis video, but I recommend you to use "Firmware_dist_calculation_fast_boot.hex".   
  
https://user-images.githubusercontent.com/10147086/146648515-e8e14f07-e56d-4fbf-a08f-5b903e57361a.mp4  

I'm not sure that it safe to power module with  5V.

**Comment by MickTheMechanic** [comment]
Thankyou for doing these tests. It is possible that my input voltage was causing problems.

**Comment by MickTheMechanic** [comment]
When you use the pc utility to measure distance, are you getting plausible measurements?

**Comment by iliasam** [comment]
Yes, I compare distance measurements with my Bosh laser tape measure.

**Comment by iliasam** [comment]
I have not tested my module with Arduino for a long time. 
Arduino sketch is published here: https://www.hackster.io/iliasam/making-a-cheap-laser-rangefinder-for-arduino-4dd849
256000 baudrate was used there.
I can't remember big problems there.
Notice, that  data speed is not very high - it is near 42*65 = 2730 bytes/s.
Yon can create nee issue au Github to report about baudrate problems.

**Comment by MickTheMechanic** [comment]
I have very good results except when the object is between 180cm and 200cm from the laser, then I have very erratic and unlausible values. Can you please confirm if you experience similar problems with your module? Even with the incresed voltage I am still have problems..

**Comment by MickTheMechanic** [comment]
I must apologise, it seems my white wallpaper has been causing the strange issues. I changed to a matt black surface and I am having good results

**Comment by bdevijve** [comment]
Hi all, I took some measures yesterday : notice VBAT is between 3.7 and 4.2V : I will probably power it from 5V here;
But PWRON is between 2.7 et 3.0V, so I will connect a GPIO from the arduino / ESP32 to put enable this device;
I will probably connect a tasmota ESP32 to it; and upload the distance to MQTT...

( I notice UART speed settings on Tasmota is between 1200 and 115200... So i will probably have to change the U85 firmware to 115200 )

**Comment by MickTheMechanic** [comment]
I have done a lot of testing and I continue to have problems with this module. I have also tested a new one and I have the same results. 

At certain distances the output becomes very erratic and implausible. If I use a Matt black surface then the problem is a bit better.

At first I thought it might be data corruption on the serial bus, but all other data is fine.

@iliasam do you experience similar problems with your module?

**Comment by iliasam** [comment]
I do several tests with my 85A module and found that results could be really unstable when amplitude is more than 700-800 points.
I would do more tests and I will try to improve code.

**Comment by maillemaker** [comment]
> I'm orded one devise here: https://www.aliexpress.com/item/1005001772064136.html I would try to investigate it.

I ordered this one to start.  It came with an 85A.

I have just ordered a second, different module:

https://www.aliexpress.com/item/1005001394260306.html

And it also came with an 85A.

**Comment by maillemaker** [comment]
Hi all,

I am following this topic with great interest.  I am a pretty new Arduino user.  It's getting hard to follow this conversation across the various sites.  

As I understand things right now

- The 85A will work, but requires an input voltage greater than 3.3V.
- The 85A can be programmed via ST-Link or UART

I have an ST-Link.

Is the pinout for the 85A the same as the 85B?

85B from tutorial: https://github.com/iliasam/Laser_tape_reverse_engineering/raw/master/PCB_photos/module_85b/pinout.jpg

My 85A: https://i.imgur.com/Z0OGLr6.jpg

**Comment by maillemaker** [comment]
It would be fantastic if this step-by-step article with pictures:

https://www.hackster.io/iliasam/making-a-cheap-laser-rangefinder-for-arduino-4dd849

Could be duplicated only using the 85A module.

**Comment by MickTheMechanic** [comment]
@maillemaker Here is the guide for how to program the module. 

https://github.com/iliasam/Laser_tape_reverse_engineering/wiki/U85-module-programming

85a can be programmed exactly the same way as the 85b module, with the same firmware. It can be done either from STLink or via serial, both methods are documented well. You can use an Arduino board as the serial connection if you don’t have a usb to serial Converter, as pins RX and TX on the Arduino board are connected directly to the inbuilt USB to Serial converter on Arduino boards.

There is currently a bug that causes erratic values when the laser intensity is too high, so until @iliasam releases new firmware the current workaround is to use the color black as your reflective surface.

**Comment by iliasam** [comment]
@MickTheMechanic Thanks for comment.
@maillemaker Currently I don't recommend to erase original firmware - because my firmware is working not stable at certain conditions, sorry for that. I'm afraid that this problem exists with 85b module too - I need time to confirm that.

The problem is that schematic of 85x modules differ somehow from previous schematic of 512/701/703 modules.
Measured phase value was always sensitive to amplitude, but newer modules have much more complicated dependence.
I don't know how it is possible, but I can see that phase dependence of amplitude is depending on "true" value of phase.
So at "true" 0 deg. measured phase is not depending on amplitude, but at "true" 90 deg measured phase is super sensitive to amplitude - phase is depending on amplitude and phase (sounds strange). So it is hard to measure phase,. 
I found this more than 4 month ago, and I implemented special compensation method than can handle with that, but you can see that this method is not working OK at certain conditions.  

All these investigations are need a lot of time, but I don't have a lot of free time, so research how to eliminate this problem could take a lot of time.  

Changing firmware with St-Link is more complicated, because you need to enable hardware reset is settings of programmer, and you programmer must have special reset pin.
 STM32CubeProgrammer can be used with St-Link, I have tested it.

**Comment by maillemaker** [comment]
> @maillemaker Here is the guide for how to program the module.
> 
> https://github.com/iliasam/Laser_tape_reverse_engineering/wiki/U85-module-programming

Yes, that is a great article for how to program the U85 module, but what is needed is an updated article on how to connect the U85A/B to the Arduino.  

> 85a can be programmed exactly the same way as the 85b module, with the same firmware. It can be done either from STLink or via serial, both methods are documented well. You can use an Arduino board as the serial connection if you don’t have a usb to serial Converter, as pins RX and TX on the Arduino board are connected directly to the inbuilt USB to Serial converter on Arduino boards.

That is good to know.  I have an STLink and Arduino Uno boards.

> There is currently a bug that causes erratic values when the laser intensity is too high, so until @iliasam releases new firmware the current workaround is to use the color black as your reflective surface.

It will be great if a new firmware is released.  In my application the laser will be operating outdoors typically against trees and foilage.  I intend to use this to "infinity" with only values within 20 - 30 feet being accepted as valid input to indicate a proximity alert.

**Comment by maillemaker** [comment]
> @MickTheMechanic Thanks for comment. @maillemaker Currently I don't recommend to erase original firmware - because my firmware is working not stable at certain conditions, sorry for that. I'm afraid that this problem exists with 85b module too - I need time to confirm that.

I thought erasing the original firmware was a precursor to installing your firmware?  I'll have to look at the instructions again.

> All these investigations are need a lot of time, but I don't have a lot of free time, so research how to eliminate this problem could take a lot of time.

I'd like to send you some PayPal cash.  Do you use PayPal?  Can I use this email?  iliasam@mail.ru

> Changing firmware with St-Link is more complicated, because you need to enable hardware reset is settings of programmer, and you programmer must have special reset pin. STM32CubeProgrammer can be used with St-Link, I have tested it.

This is the ST-Link I have purchased:

https://smile.amazon.com/gp/product/B07SQV6VLZ

It appears to have pins for:

RST
SWDIO
GND
GND
SWIM
SWCLOCK
3.3V
3.3V
5.0V
5.0V

I assume RST is the reset pin needed.

![image](https://user-images.githubusercontent.com/26260929/147698604-1c3827ed-647d-4523-9758-903ca871b8ae.png)

**Comment by iliasam** [comment]
"is needed is an updated article on how to connect the U85A/B to the Arduino."
Just connect RX and TX pin of 85x module to Arduino, as it is done for previous modules. Notice that working with keyboard is not  implemented now, so calibration command can be send only by UART.

"I thought erasing the original firmware was a precursor to installing your firmware?" You need to erase flash with original firmware to remove flash lock. Current firmware is not stable, so I do not recommend to change firmware at all - because there is no way back.   
  
"I'd like to send you some PayPal cash" Thanks, but I don't need that.
I'm not sure, that I can complete this project at all.

"This is the ST-Link I have purchased"
It is Chinese clone, its reset pin for STM8 only: https://ziutek.github.io/2018/03/26/stlink.html

**Comment by maillemaker** [comment]
Sounds like my ST-Link is out then.  Guess I'll try using the Arduino to talk to it.

I was really excited about this project because I need a laser tripwire mechanism.  There are other simple implementations but they generally require a reflector at the other end, and so make a "beam breaking" mechanism.  There are also IR motion sensors but they tend not to be as tight a focus nor as long a range as a tripwire.

I figured the laser rangefinder solution was ideal.  Simply set the device and then monitor if the range ever reads less than 30 feet (or whatever) and if it does, trigger.

Wish someone made a plug-and-play solution for Arduino.

**Comment by iliasam** [comment]
@maillemaker Maybe this would be interesting for you: https://aliexpress.ru/item/1005003256773059.html
This module is sold with original firmware with UART enabled.

**Comment by maillemaker** [comment]
> @maillemaker Maybe this would be interesting for you: https://aliexpress.ru/item/1005003256773059.html This module is sold with original firmware with UART enabled.

Thanks iliasam.  I've been doing some googling last night and this morning and I have found some modules like you found.  But they are $40-$70 or so.  

Here is another:

https://www.dfrobot.com/product-2108.html

Here is something that may be promising in the $25 range:

https://www.amazon.com/SmartFly-info-TF-Luna-Short-Range-Single-Point/dp/B08F55QTWP/ref=psdc_7459508011_t3_B07XXTMRR2

But I understand these time-of-flight sensors are too fast for an Arduino to process and so require some special high-speed hardware to capture the information.  Not sure though I'm just now learning about them.

Anyway I wouldn't mind spending $100 on a single project piece but I'm hoping to build these into a toy airsoft mine and so I was really excited by the $12 tape measure solution.

I'm going to try to contact JRT again and see if they would be interested in making a low-cost stand-along Arduino module.  IF they can sell them to tape measure companies for a couple of bucks they should be able to sell them to consumers.

**Comment by MickTheMechanic** [comment]
@maillemaker I have already been in touch with JRT Measure, they offer a module with UART enabled that is suitable for use with Arduino. They quoted me US$49 for the module and US$25 postage to Germany.

you should take a look at the VLX53 module. It is an inexpensive TOF sensor that is easy to use with Arduino. I was previously using it for my project, but the accuracy is +/- 10mm so it wasn’t accurate enough for me. But it would be very easy to build the project you describe with this sensor. It cost only approx €20

**Comment by MickTheMechanic** [comment]
@iliasam I am very sorry to hear you are struggling with the project and that you might not be able to complete the new firmware. 

I have had some very good results with your current firmware, I ran the results through a Kalman filter and used a Matt black surface, and the results were very accurate, only at some distances the results became erratic, but it was still very encouraging. Because of your firmware I was able to build a working prototype of a wheel alignment device I designed.

I suppose I should ask, is it worth waiting for new firmware, or should I instead search for some different sensors?

**Comment by iliasam** [comment]
@maillemaker The module that I mentioned is made by JRM Measure.
As I can understand they are slow enough and can be easily used with Arduino.

TF-Luna is also can be used with Arduino: https://www.youtube.com/watch?v=5AFgg6C7hxg

@MickTheMechanic I'm not sure about needed time to improve firmware. It can take several month.

**Comment by maillemaker** [comment]
> @maillemaker I have already been in touch with JRT Measure, they offer a module with UART enabled that is suitable for use with Arduino. They quoted me US$49 for the module and US$25 postage to Germany.

That's would be OK for a prototype/project, but not feasible if I wanted to try and sell the airsoft mine.  There were some "claymore" style of mines made with dual IR sensors, but they were known to be quirky/unreliable to go off.  I have been experimenting with the IR motion sensor anyway and it does seem to work but it has quite a wide cone of vision it may not be suitable.  I was going to try and limit it's cone of vision by making it peek through an aperture but I understand the device works through differential detection across the field of view so limiting its field of view may reduce its detection capabilities.  

It's a shame if they can produce them for such low cost applications as the digital tape measures that they cannot offer something for hobbyists for $10 or so.

> you should take a look at the VL53X module. It is an inexpensive TOF sensor that is easy to use with Arduino. I was previously using it for my project, but the accuracy is +/- 10mm so it wasn’t accurate enough for me. But it would be very easy to build the project you describe with this sensor. It cost only approx €20

Looks like it only has a range of a couple of meters.  I think for my application I need at least 8 meters.  10 meters would be better.

**Comment by maillemaker** [comment]
> @maillemaker The module that I mentioned is made by JRM Measure. As I can understand they are slow enough and can be easily used with Arduino.
> 
> TF-Luna is also can be used with Arduino: https://www.youtube.com/watch?v=5AFgg6C7hxg
> 
> @MickTheMechanic I'm not sure about needed time to improve firmware. It can take several month.

I assume you mean JRT Measure, not JRM.

I just discovered the TF-Lulna  yesterday after renewing my google searches for "arduino laser rangefinder".  It appears to have a range of 8m so that will just work.  I will give it a try.

The nice thing about the TF-Luna and the VL53X is they are infrared so there will be no tell-tale red dot somewhere that someone might notice and thus avoid the mine!  :)

**Comment by bdevijve** [comment]
In the meantime, I'm trying to get the firmware running UART from JRT...
I'll tell you...

Benoît

**Comment by iliasam** [comment]
@bdevijve 
JRT is selling modules with UART support for 30-40$ while laser tape measure is sold for 12-15$. Original firmware is locked in MCU.
Do you believe that they will really give you firmware?

**Comment by bdevijve** [comment]
Yes I know this... But let's give it a chance !

**Comment by iliasam** [comment]
It looks like I have solved this problem. Now module is much less sensitive to amplitude changings.
It would be be great if someone here would test this new firmware
[Firmware_dist_calculation_fast_boot_115200.zip](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7865289/Firmware_dist_calculation_fast_boot_115200.zip)
You can see that this firmware is working at 115200 baudrate.

This utility can be used for testing:
[PC_show_results_v1_1.zip](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7865300/PC_show_results_v1_1.zip)

**Comment by bdevijve** [comment]
I will test very soon,
Can you confirm the range to test and the reflector to use for "worst case" testing ?

**Comment by MickTheMechanic** [comment]
I will also test soon.

Can I ask what was causing the problem with higher amplitudes?

**Comment by bdevijve** [comment]
Hi all,

I flashed my device and make some tests between my desk and a wall at 16m from me,

1/ I don't know what your program is recording but I have far less measurements the file that the RX points count...
(in file Result_2022-01-14.02-30-06.txt, RX points was 5000 but the file is 800 lines long)

[Result_2022-01-14 02-30-06.txt](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7867007/Result_2022-01-14.02-30-06.txt)

2/ I that same file, I notice a "lower distance value" every 15 samples or so...

3/ I made the following move with a paper in my hand
* Logs are in the file "Result_2022-01-14.02-20-56.txt"
[Result_2022-01-14 02-20-56.txt](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7867020/Result_2022-01-14.02-20-56.txt)
A/ I moved away with a dark paper in my hand, starting from very close and ending at 16m about
B/ Then I moved back with a white paper

I notice on A/ something like 10% are wrong when reaching 9m-16m; this issue is not seeing on B/

Let us know if everything is as expected or not; but from my point of view it is far from enough to do what I need !!

Many thanks !

**Comment by bdevijve** [comment]
I did a second time the test 2/
-> The raw values out from UART are better:
![image](https://user-images.githubusercontent.com/74436690/149438530-e2a46667-693c-40d5-af15-aee5cdaa9ca5.png)
[putty.log](https://github.com/iliasam/Laser_tape_reverse_engineering/files/7867110/putty.log)

**Comment by bdevijve** [comment]
![MVIMG_20220114_030513](https://user-images.githubusercontent.com/74436690/149439116-80f71430-baab-4b87-88b0-3409b5c0af87.jpg)
![MVIMG_20220114_030442](https://user-images.githubusercontent.com/74436690/149439120-854e5ddb-eea4-4146-9795-4504f8dbf709.jpg)

**Comment by iliasam** [comment]
@MickTheMechanic The problem was with ADC settings. Firstly I set too low sampling time, witch seems for me produce a kind of interference from ADC measurement channel to ADC reference channel. And this interference was depending of signal amplitude.

@bdevijve Thanks for your measurements.
It is predictable that at big distances and with dark object you have very low signal amplitude: 10-40 ADC points, which is close to noise. So distance is measured really bad. 
Original firmware is doing a lot of measurements and a kind of averaging them to get stable result. Measurement cycle duration can take 2s or more. 
I tried to do median filtering in Excel of your "putty.log" data. I need to use 40 points filtering to remove all bad results:
![image](https://user-images.githubusercontent.com/10147086/149485188-ee90272b-fc18-421f-8fe9-c90a2cb46ad4.png)

**Comment by bdevijve** [comment]
This is almost perfect for my use case I think...
I want to meter the water level in a well - with distances between 1.5m and 4m
Reflexion surface is going to be the water, and environment is 99% dark all time...

Anyway getting a valid result every 1s after performing some maths is way fast enough for me

I'll make this final test soon...

Benoît

**Comment by MickTheMechanic** [comment]
@iliasam so far I am having good results. I will continue testing and let you know if any further problems arise.

**Comment by iliasam** [comment]
I have updated firmware - changed ADC settings to get best temperature stability.
I have tested module in climatic chamber with new firmware, distance measurement dependence should be less that 5 mm.
It would be good if someone will test it.
See HEX here: https://github.com/iliasam/Laser_tape_reverse_engineering/tree/master/Code/CortexM0

**Comment by MickTheMechanic** [comment]
Hi, sorry for the late reply. I am exited to test the new firmware, but I have other projects at the moment and I cannot test until the und of the month. I will keep in touch and let you know the results from testing.

Appreciate your work :)

**Comment by bdevijve** [comment]
Ok I will update it today on me side and confirm all is right !

**Comment by bdevijve** [comment]
Just updated: still working perfectly :)

**Comment by MickTheMechanic** [comment]
Hi, I have recently revisit my old project and can confirm the new firmware is also working for me. Thanks a lot for your work, really appreciate it.
